include/sys/notify.h

```diff
@@ -0,0 +1,361 @@
+ *
+ * Objects of this type are allocated by a client, which must use an
+ * initialization function (e.g. sys_notify_init_signal()) to
+ * configure them.  Generally the structure is a member of a
+ * service-specific client structure, such as onoff_client.
+ *
+ * Control of the containing object transfers to the service provider
+ * when a pointer to the object is passed to a service function that
+ * is documented to take control of the object, such as
+ * onoff_service_request().  While the service provider controls the
+ * object the client must not change any object fields.  Control
+ * reverts to the client:
+ * * if the call to the service API returns an error;
+ * * when operation completion is posted.  This may occur before the
+ *   call to the service API returns.
+ *
+ * Operation completion is technically posted when the flags field is
+ * updated so that sys_notify_fetch_result() returns success.  This
+ * will happen before the signal is posted or callback is invoked.
+ * Note that although the manager will no longer reference the
+ * sys_notify object past this point, the containing object may have
+ * state that will be referenced within the callback.  Where callbacks
+ * are used control of the containing object does not revert to the
+ * client until the callback has been invoked.  (Re-use within the
+ * callback is explicitly permitted.)
+ *
+ * After control has reverted to the client the notify object must be
+ * reinitialized for the next operation.
+ *
+ * The content of this structure is not public API to clients: all
+ * configuration and inspection should be done with functions like
+ * sys_notify_init_callback() and sys_notify_fetch_result().
+ * However, services that use this structure may access certain
+ * fields directly.
+ */
+struct sys_notify {
+	union method {
+		/* Pointer to signal used to notify client.
+		 *
+		 * The signal value corresponds to the res parameter
+		 * of sys_notify_callback.
+		 */
+		struct k_poll_signal *signal;
+
+		/* Generic callback function for callback notification. */
+		sys_notify_generic_callback callback;
+	} method;
+
+	/*
+	 * Flags recording information about the operation.
+	 *
+	 * Bits below SYS_NOTIFY_EXTENSION_POS are initialized by
+	 * async notify API init functions like
+	 * sys_notify_init_callback(), and must not by modified by
+	 * extensions or client code.
+	 *
+	 * Bits at and above SYS_NOTIFY_EXTENSION_POS are available
+	 * for use by service extensions while the containing object
+	 * is managed by the service.  They are not for client use,
+	 * are zeroed by the async notify API init functions, and will
+	 * be zeroed by sys_notify_finalize().
+	 */
+	u32_t volatile flags;
+
+	/*
+	 * The result of the operation.
+	 *
+	 * This is the value that was (or would be) passed to the
+	 * async infrastructure.  This field is the sole record of
+	 * success or failure for spin-wait synchronous operations.
+	 */
+	int volatile result;
+};
+
+/** @internal */
```

Comment:
In other parts of the code, I have used
```
/** @cond INTERNAL_HIDDEN */
...
/** @endcond */
```
not sure which is correct

Reply:
I'm not sure either work, but `@internal` is common.